### PR TITLE
immutable Instruction

### PR DESCRIPTION
**Pull request by andreaTP**
Not sure if this is the final API we are aiming for, but, at least, this goes in the direction of having only immutable classes in the `wasm` module.

### CODE REVIEW

File: aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java [hunk 1]
```diff
@@ -688,10 +692,10 @@ private void compileBody(
                     break;
                 case BLOCK:
                 case LOOP:
-                    ctx.enterScope(ins.scope(), blockType(ins));
+                    ctx.enterScope(ins.scope().get(), blockType((int) ins.operands()[0]));
```
Comment (electrum) [inline aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java:695]: Using `Optional` for these fields hasn't improved type safety much, as the caller is unconditionally getting the values. The advantage is that it will fail immediately, rather than providing a null. What we don't have is enforcement between the opcode type and the type-specific fields.

We could improve this by making these methods return the values directly, but throw `IllegalStateException`  if the opcode type is not valid. At construction time, we can enforce that the metadata values are set if and only if the opcode type matches.

In other words, `scope()` must be set for, and can only be set for and called for, the `BLOCK`, `LOOP`, `END`, and `IF` types. Similar logic applies to the other fields.

**Review by electrum — COMMENTED** [review]
I like where this is going, but I want to think about it more.

What do you think about having `Instruction` with the three core fields and having an `AnnotatedInstruction` subclass with the metadata fields?

**Comment by andreaTP** [comment]
@electrum great suggestions, I'll apply everything and we iterate :+1:

**Comment by andreaTP** [comment]
Renaming `Instruction` to `AnnotatedInstruction` caused the diff to be a bit messy, I have no strong opinions on the names here, so I'm happy to accept an alternative suggestion.

In general, I think that @electrum feedback has been valuable and I like how the code looks after this iteration.
Should be ready for the next round of review :nerd_face:

File: aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java [hunk 2]
```diff
@@ -642,17 +648,15 @@ private void compileBody(
...
-            if (ins.labelTable() != null) {
-                for (int label : ins.labelTable()) {
-                    labels.put(label, new Label());
-                }
-            }
+            ins.labelTrue().ifPresent(l -> labels.put(l, new Label()));
+            ins.labelFalse().ifPresent(l -> labels.put(l, new Label()));
+            ins.labelTable()
+                    .ifPresent(
+                            l -> {
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java]: removed the `Optional`

File: aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java [hunk 3]
```diff
@@ -79,7 +80,12 @@
 public final class AotMachine implements Machine {
 
     public static final String DEFAULT_CLASS_NAME = "com.dylibso.chicory.$gen.CompiledModule";
-    private static final Instruction FUNCTION_SCOPE = new Instruction(-1, OpCode.NOP, new long[0]);
+    private static final BaseInstruction FUNCTION_SCOPE =
+            BaseInstruction.builder()
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java]: done

File: aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java (hunk 1 above)
Comment (andreaTP) [reply aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java]: done

File: wasm/src/main/java/com/dylibso/chicory/wasm/types/BaseInstruction.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
+package com.dylibso.chicory.wasm.types;
+
+public class BaseInstruction {
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/types/BaseInstruction.java]: done

File: aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java [hunk 5]
```diff
@@ -642,17 +648,15 @@ private void compileBody(
...
-            }
-            if (ins.labelFalse() != null) {
-                labels.put(ins.labelFalse(), new Label());
-            }
-            if (ins.labelTable() != null) {
-                for (int label : ins.labelTable()) {
-                    labels.put(label, new Label());
-                }
-            }
+            ins.labelTrue().ifPresent(l -> labels.put(l, new Label()));
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply aot/src/main/java/com/dylibso/chicory/aot/AotMachine.java]: removed all the `Optional`s, I'll keep it in mind going forward.

File: wasm/src/main/java/com/dylibso/chicory/wasm/types/Instruction.java [hunk 6]
```diff
@@ -6,19 +6,17 @@ public class Instruction {
     private final int address;
     private final OpCode opcode;
     private final long[] operands;
```
Comment (electrum) [inline wasm/src/main/java/com/dylibso/chicory/wasm/types/Instruction.java:8]: As a follow up, we need to fix this, as exposing an array means it is not immutable. To avoid boxing, we could keep it as an array internally, but add accessors `operand0()` and `operand1()`. The `operands()` method would return a clone of the array.

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
The diff does look messier, but this feels like a cleaner API.

File: wasm/src/main/java/com/dylibso/chicory/wasm/types/AnnotatedInstruction.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            return this;
+        }
+
+        public AnnotatedInstruction build() {
+            switch (base.opcode()) {
+                case BLOCK:
+                case LOOP:
+                case END:
+                case IF:
+                    assert (scope.isPresent());
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/types/AnnotatedInstruction.java:137]: > Do we enable assertions when running unit tests?

yes

> is the assumption that these are only needed when developing Chicory and they should be off at runtime?

this is accurate, to the best of my understanding

File: wasm/src/main/java/com/dylibso/chicory/wasm/types/Instruction.java (hunk 6 above)
Comment (andreaTP) [reply wasm/src/main/java/com/dylibso/chicory/wasm/types/Instruction.java:8]: make sense, let's follow up :+1:

**Review by electrum — APPROVED** [review]
